man/paplay.1
.TH paplay 1 User Manuals
.SH NAME
paplay \- Play back audio files on a PulseAudio sound server
.SH SYNOPSIS
\fBpaplay [\fIoptions\fB] \fIFILE\fB

paplay --help\fB

paplay --version\fB
\f1
.SH DESCRIPTION
\fIpaplay\f1 is a simple tool for playing back audio files on a PulseAudio sound server. It understands all audio file formats supported by \fIlibsndfile\f1.
.SH OPTIONS
.TP
\fB-h | --help\f1
Show help.
.TP
\fB--version\f1
Show version information.
.TP
\fB-v | --verbose\f1
Enable verbose operation. Dumps the current playback time to STDERR during playback.
.TP
\fB-s | --server\f1\fI=SERVER\f1
Choose the server to connect to.
.TP
\fB-d | --device\f1\fI=SINK\f1
Specify the symbolic name of the sink to play this file on.
.TP
\fB-n | --client-name\f1\fI=NAME\f1
Specify the client name \fIpaplay\f1 shall pass to the server when connecting.
.TP
\fB--stream-name\f1\fI=NAME\f1
Specify the stream name \fIpaplay\f1 shall pass to the server when creating the stream.
.TP
\fB--volume\f1\fI=VOLUME\f1
Specify the initial playback volume to use. Choose a value between 0 (silent) and 65536 (100% volume).
.TP
\fB--channel-map\f1\fI=CHANNELMAP\f1
Explicitly choose a channel map when playing back this stream. The argument should be a comma separated list of channel names: \fIfront-left\f1, \fIfront-right\f1, \fImono\f1, \fIfront-center\f1, \fIrear-left\f1, \fIrear-right\f1, \fIrear-center\f1, \fIlfe\f1, \fIfront-left-of-center\f1, \fIfront-right-of-center\f1, \fIside-left\f1, \fIside-right\f1, \fItop-center\f1, \fItop-front-center\f1, \fItop-front-left\f1, \fItop-front-right\f1, \fItop-rear-left\f1, \fItop-rear-right\f1, \fItop-rear-center\f1, or any of the 32 auxiliary channel names \fIaux0\f1 to \fIaux31\f1.
.SH LIMITATIONS
Due to a limitation in \fIlibsndfile\f1 \fIpaplay\f1 currently does not always set the correct channel mapping for playback of multichannel (i.e. surround) audio files, even if the channel mapping information is available in the audio file.
.SH AUTHORS
The PulseAudio Developers <mzchyfrnhqvb (at) 0pointer (dot) net>; PulseAudio is available from \fBhttp://pulseaudio.org/\f1
.SH SEE ALSO
\fBpulseaudio(1)\f1, \fBpacat(1)\f1


ChangeLog
2007-11-21 23:09  lennart

	* configure.ac: bump version and soname

2007-11-21 23:03  lennart

	* man/pacat.1.xml.in, man/pulse-daemon.conf.5.xml.in: update man
	  pages a bit

2007-11-21 22:55  lennart

	* src/pulsecore/protocol-native.c: when speaking to a client with a
	  version < 12, hide S32 sample specs, and make them appaear as
	  FLOAT32

2007-11-21 01:31  lennart

	* configure.ac: increment api and protocol version

2007-11-21 01:30  lennart

	* src/daemon/daemon-conf.c, src/daemon/daemon-conf.h,
	  src/daemon/daemon.conf.in, src/daemon/main.c, src/map-file,
	  src/modules/module-suspend-on-idle.c,
	  src/modules/module-volume-restore.c, src/pulse/context.c,
	  src/pulse/def.h, src/pulse/internal.h, src/pulse/operation.c,
	  src/pulse/stream.c, src/pulse/stream.h, src/pulsecore/cli-text.c,
	  src/pulsecore/core.c, src/pulsecore/core.h, src/pulsecore/pid.c,
	  src/pulsecore/pid.h, src/pulsecore/protocol-native.c,
	  src/pulsecore/sink-input.c, src/pulsecore/sink-input.h,
	  src/pulsecore/sink.c, src/pulsecore/source-output.c,
	  src/pulsecore/source-output.h, src/pulsecore/source.c,
	  src/utils/pacat.c, src/utils/pacmd.c: - Check process name when
	  dealing with PID files
	  - Add new PA_STREAM_FIX_CHANNELS, FIX_RATE, FIX_FORMAT,
	  DONT_MOVE, VARIABLE_RATES to pa_sream_flags_t adn implement it
	  - Expose those flags in pacat
	  - Add notifications about device suspend/resume to the protocol
	  and expose them in libpulse
	  - Allow changing of buffer_attr during playback
	  - allow disabling for remixing globally
	  - hookup polkit support

2007-11-21 01:21  lennart

	* src/modules/module-oss.c: minor typo fix

2007-11-21 01:21  lennart

	* PROTOCOL: add short version history of the PA protocol

2007-11-21 01:20  lennart

	* src/pulsecore/native-common.h: add a couple of new opcodes, and
	  document the versions the opcodes where added

2007-11-21 01:19  lennart

	* src/pulsecore/memblockq.c, src/pulsecore/memblockq.h: add API to
	  allow runtime reconfiguration of memblockqs

2007-11-16 23:48  lennart

	* configure.ac: detect whether PolicyKit support is available

2007-11-16 23:47  lennart

	* src/Makefile.am: build PolicyKit support

2007-11-16 23:46  lennart

	* src/daemon/PulseAudio.policy, src/daemon/polkit.c,
	  src/daemon/polkit.h: add interface to PolicyKit

2007-11-14 16:11  lennart

	* src/pulse/thread-mainloop.c: use a prio inheriting mutex for the
	  threaded mainloop, to ease writing of RT clients

2007-11-14 16:11  lennart

	* src/pulse/operation.c: use a free list for allocation
	  pa_operation objects

2007-11-14 16:10  lennart

	* src/pulsecore/pdispatch.c: use a free list for allocating
	  reply_info structs

2007-11-14 16:09  lennart

	* src/modules/module-jack-sink.c, src/modules/module-jack-source.c:
	  add missing pa_boolization

2007-11-13 23:42  lennart

	* src/modules/alsa-util.c, src/modules/alsa-util.h,
	  src/modules/module-alsa-sink.c, src/modules/module-alsa-source.c:
	  Rework ALSA mixer channel detection code. This time we actually
	  care about the channel names the ALSA mixer exports for us

2007-11-13 23:41  lennart

	* src/pulse/channelmap.c: add array size to increase chance of
	  detecting missing updates

2007-11-13 19:56  lennart

	* src/Makefile.am: don't fail if the bt-proximity-helper is not
	  built

2007-11-13 19:39  lennart

	* src/daemon/daemon-conf.c, src/daemon/daemon.conf.in: don't touch
	  RLIMIT:MEMBLOCK by default. This should improve out-of-the-box
	  comaptibility with JACK

2007-11-13 19:21  lennart

	* src/utils/pacmd.c: increase the pacmd timeout a bit

2007-11-13 17:37  lennart

	* src/modules/alsa-util.c, src/modules/alsa-util.h,
	  src/modules/module-alsa-sink.c, src/modules/module-alsa-source.c,
	  src/modules/module-bt-proximity.c, src/modules/module-cli.c,
	  src/modules/module-detect.c, src/modules/module-hal-detect.c,
	  src/modules/module-oss.c, src/modules/rtp/module-rtp-send.c,
	  src/pulsecore/modargs.c, src/pulsecore/modargs.h,
	  src/pulsecore/protocol-esound.c, src/pulsecore/protocol-native.c,
	  src/pulsecore/protocol-simple.c: Completely rework ALSA device
	  selection code: choose the device to open depending on the
	  requested number of channels and channel map. In most cases it
	  will now suffice to set default-channels=6 to enable 5.1 sound
	  for all devices that support it

2007-11-13 17:35  lennart

	* src/pulsecore/module.c: fix loading of load-once modules if no
	  other modules was loaded before

2007-11-11 23:18  lennart

	* configure.ac, src, src/Makefile.am,
	  src/modules/bt-proximity-helper.c,
	  src/modules/module-bt-proximity.c: add new fun module that
	  automatically mutes your audio devices when you leave with your
	  bluetooth phone, and unmutes when you come back

2007-11-11 23:01  lennart

	* src/modules/gconf/module-gconf.c: Port module-gconf to make use
	  of the new API pa_start_child_for_read()

2007-11-11 23:00  lennart

	* src/pulsecore/start-child.c, src/pulsecore/start-child.h: pull
	  code for starting helper processes out of module-gconf, clean it
	  up, and stick into a new API pa_start_child_for_read()

2007-11-11 22:59  lennart

	* src/pulsecore/macro.h: add new function pa_strnull() to simplify
	  passing null strings to non-linux printf()

2007-11-11 02:30  lennart

	* src, src/Makefile.am, src/pulsecore/resampler.c,
	  src/pulsecore/resampler.h, src/pulsecore/sink-input.c,
	  src/pulsecore/source-output.c, src/tests/remix-test.c,
	  src/tests/resampler-test.c: add a simple fully-automatic
	  fully-linearupmixer/downmixer and enable it by default

2007-11-09 18:25  lennart

	* src/daemon/dumpmodules.c, src/modules/gconf/module-gconf.c,
	  src/modules/module-alsa-sink.c, src/modules/module-alsa-source.c,
	  src/modules/module-cli.c, src/modules/module-combine.c,
	  src/modules/module-default-device-restore.c,
	  src/modules/module-defs.h.m4, src/modules/module-detect.c,
	  src/modules/module-esound-compat-spawnfd.c,

src/utils/paplay.c
/* $Id: paplay.c 2025 2007-11-05 23:56:30Z lennart $ */

/***
  This file is part of PulseAudio.

  Copyright 2004-2006 Lennart Poettering
  Copyright 2006 Pierre Ossman <ossman@cendio.se> for Cendio AB

  PulseAudio is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published
  by the Free Software Foundation; either version 2 of the License,
  or (at your option) any later version.

  PulseAudio is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with PulseAudio; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA.
***/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <signal.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <locale.h>

#include <sndfile.h>

#include <pulse/pulseaudio.h>

#if PA_API_VERSION < 9
#error Invalid PulseAudio API version
#endif

static pa_context *context = NULL;
static pa_stream *stream = NULL;
static pa_mainloop_api *mainloop_api = NULL;

static char *stream_name = NULL, *client_name = NULL, *device = NULL;

static int verbose = 0;
static pa_volume_t volume = PA_VOLUME_NORM;

static SNDFILE* sndfile = NULL;

static pa_sample_spec sample_spec = { 0, 0, 0 };
static pa_channel_map channel_map;
static int channel_map_set = 0;

static sf_count_t (*readf_function)(SNDFILE *_sndfile, void *ptr, sf_count_t frames) = NULL;

/* A shortcut for terminating the application */
static void quit(int ret) {
    assert(mainloop_api);
    mainloop_api->quit(mainloop_api, ret);
}

/* Connection draining complete */
static void context_drain_complete(pa_context*c, void *userdata) {
    pa_context_disconnect(c);
}

/* Stream draining complete */
static void stream_drain_complete(pa_stream*s, int success, void *userdata) {
    pa_operation *o;

    if (!success) {
        fprintf(stderr, "Failed to drain stream: %s\n", pa_strerror(pa_context_errno(context)));
        quit(1);
    }

    if (verbose)
        fprintf(stderr, "Playback stream drained.\n");

    pa_stream_disconnect(stream);
    pa_stream_unref(stream);
    stream = NULL;

    if (!(o = pa_context_drain(context, context_drain_complete, NULL)))
        pa_context_disconnect(context);
    else {
        pa_operation_unref(o);

        if (verbose)
            fprintf(stderr, "Draining connection to server.\n");
    }
}

/* This is called whenever new data may be written to the stream */
static void stream_write_callback(pa_stream *s, size_t length, void *userdata) {
    sf_count_t bytes;
    void *data;
    assert(s && length);

    if (!sndfile)
        return;

    data = pa_xmalloc(length);

    if (readf_function) {
        size_t k = pa_frame_size(&sample_spec);

        if ((bytes = readf_function(sndfile, data, length/k)) > 0)
            bytes *= k;

    } else
        bytes = sf_read_raw(sndfile, data, length);

    if (bytes > 0)
        pa_stream_write(s, data, bytes, pa_xfree, 0, PA_SEEK_RELATIVE);
    else
        pa_xfree(data);

    if (bytes < (sf_count_t) length) {
        sf_close(sndfile);
        sndfile = NULL;
        pa_operation_unref(pa_stream_drain(s, stream_drain_complete, NULL));
    }
}

/* This routine is called whenever the stream state changes */
static void stream_state_callback(pa_stream *s, void *userdata) {
    assert(s);

    switch (pa_stream_get_state(s)) {
        case PA_STREAM_CREATING:
        case PA_STREAM_TERMINATED:
            break;

        case PA_STREAM_READY:
            if (verbose)
                fprintf(stderr, "Stream successfully created\n");
            break;

        case PA_STREAM_FAILED:
        default:
            fprintf(stderr, "Stream errror: %s\n", pa_strerror(pa_context_errno(pa_stream_get_context(s))));
            quit(1);
    }
}

/* This is called whenever the context status changes */
static void context_state_callback(pa_context *c, void *userdata) {
    assert(c);

    switch (pa_context_get_state(c)) {
        case PA_CONTEXT_CONNECTING:
        case PA_CONTEXT_AUTHORIZING:
        case PA_CONTEXT_SETTING_NAME:
            break;

        case PA_CONTEXT_READY: {
            pa_cvolume cv;

            assert(c && !stream);

            if (verbose)
                fprintf(stderr, "Connection established.\n");

            stream = pa_stream_new(c, stream_name, &sample_spec, channel_map_set ? &channel_map : NULL);
            assert(stream);

            pa_stream_set_state_callback(stream, stream_state_callback, NULL);
            pa_stream_set_write_callback(stream, stream_write_callback, NULL);
            pa_stream_connect_playback(stream, device, NULL, 0, pa_cvolume_set(&cv, sample_spec.channels, volume), NULL);

            break;
        }

        case PA_CONTEXT_TERMINATED:
            quit(0);
            break;

        case PA_CONTEXT_FAILED:
        default:
            fprintf(stderr, "Connection failure: %s\n", pa_strerror(pa_context_errno(c)));
            quit(1);
    }
}

/* UNIX signal to quit recieved */
static void exit_signal_callback(pa_mainloop_api*m, pa_signal_event *e, int sig, void *userdata) {
    if (verbose)
        fprintf(stderr, "Got SIGINT, exiting.\n");
    quit(0);

}

static void help(const char *argv0) {

    printf("%s [options] [FILE]\n\n"
           "  -h, --help                            Show this help\n"
           "      --version                         Show version\n\n"
           "  -v, --verbose                         Enable verbose operation\n\n"
           "  -s, --server=SERVER                   The name of the server to connect to\n"
           "  -d, --device=DEVICE                   The name of the sink to connect to\n"
           "  -n, --client-name=NAME                How to call this client on the server\n"
           "      --stream-name=NAME                How to call this stream on the server\n"
           "      --volume=VOLUME                   Specify the initial (linear) volume in range 0...65536\n"
           "      --channel-map=CHANNELMAP          Set the channel map to the use\n",
           argv0);
}

enum {
    ARG_VERSION = 256,
    ARG_STREAM_NAME,
    ARG_VOLUME,
    ARG_CHANNELMAP
};

int main(int argc, char *argv[]) {
    pa_mainloop* m = NULL;
    int ret = 1, r, c;
    char *bn, *server = NULL;
    const char *filename;
    SF_INFO sfinfo;

    static const struct option long_options[] = {
        {"device",      1, NULL, 'd'},
        {"server",      1, NULL, 's'},
        {"client-name", 1, NULL, 'n'},
        {"stream-name", 1, NULL, ARG_STREAM_NAME},
        {"version",     0, NULL, ARG_VERSION},
        {"help",        0, NULL, 'h'},
        {"verbose",     0, NULL, 'v'},
        {"volume",      1, NULL, ARG_VOLUME},
        {"channel-map", 1, NULL, ARG_CHANNELMAP},
        {NULL,          0, NULL, 0}
    };

    setlocale(LC_ALL, "");

    if (!(bn = strrchr(argv[0], '/')))
        bn = argv[0];
    else
        bn++;

    while ((c = getopt_long(argc, argv, "d:s:n:hv", long_options, NULL)) != -1) {

        switch (c) {
            case 'h' :
                help(bn);
                ret = 0;
                goto quit;

            case ARG_VERSION:
                printf("paplay "PACKAGE_VERSION"\nCompiled with libpulse %s\nLinked with libpulse %s\n", pa_get_headers_version(), pa_get_library_version());
                ret = 0;
                goto quit;

            case 'd':
                pa_xfree(device);
                device = pa_xstrdup(optarg);
                break;

            case 's':
                pa_xfree(server);
                server = pa_xstrdup(optarg);
                break;

            case 'n':
                pa_xfree(client_name);
                client_name = pa_xstrdup(optarg);
                break;

            case ARG_STREAM_NAME:
                pa_xfree(stream_name);
                stream_name = pa_xstrdup(optarg);
                break;

            case 'v':
                verbose = 1;
                break;

            case ARG_VOLUME: {
                int v = atoi(optarg);
                volume = v < 0 ? 0 : v;
                break;
            }

            case ARG_CHANNELMAP:
                if (!pa_channel_map_parse(&channel_map, optarg)) {
                    fprintf(stderr, "Invalid channel map\n");
                    goto quit;
                }

                channel_map_set = 1;
                break;

            default:
                goto quit;
        }
    }

    filename = optind < argc ? argv[optind] : "STDIN";

    memset(&sfinfo, 0, sizeof(sfinfo));

    if (optind < argc)
        sndfile = sf_open(filename, SFM_READ, &sfinfo);
    else
        sndfile = sf_open_fd(STDIN_FILENO, SFM_READ, &sfinfo, 0);

    if (!sndfile) {
        fprintf(stderr, "Failed to open file '%s'\n", filename);
        goto quit;
    }

    sample_spec.rate = sfinfo.samplerate;
    sample_spec.channels = sfinfo.channels;

    readf_function = NULL;

    switch (sfinfo.format & 0xFF) {
        case SF_FORMAT_PCM_16:
        case SF_FORMAT_PCM_U8:
        case SF_FORMAT_PCM_S8:
            sample_spec.format = PA_SAMPLE_S16NE;
            readf_function = (sf_count_t (*)(SNDFILE *_sndfile, void *ptr, sf_count_t frames)) sf_readf_short;
            break;

        case SF_FORMAT_ULAW:
            sample_spec.format = PA_SAMPLE_ULAW;
            break;

        case SF_FORMAT_ALAW:
            sample_spec.format = PA_SAMPLE_ALAW;
            break;

        case SF_FORMAT_FLOAT:
        case SF_FORMAT_DOUBLE:
        default:
            sample_spec.format = PA_SAMPLE_FLOAT32NE;
            readf_function = (sf_count_t (*)(SNDFILE *_sndfile, void *ptr, sf_count_t frames)) sf_readf_float;
            break;
    }

    assert(pa_sample_spec_valid(&sample_spec));

    if (channel_map_set && channel_map.channels != sample_spec.channels) {
        fprintf(stderr, "Channel map doesn't match file.\n");
        goto quit;
    }

    if (!client_name) {
        client_name = pa_locale_to_utf8(bn);
        if (!client_name)
            client_name = pa_utf8_filter(bn);
    }

    if (!stream_name) {
        const char *n;

        n = sf_get_string(sndfile, SF_STR_TITLE);

        if (!n)
            n = filename;

        stream_name = pa_locale_to_utf8(n);
        if (!stream_name)
            stream_name = pa_utf8_filter(n);
    }

    if (verbose) {
        char t[PA_SAMPLE_SPEC_SNPRINT_MAX];
        pa_sample_spec_snprint(t, sizeof(t), &sample_spec);
        fprintf(stderr, "Using sample spec '%s'\n", t);
    }

    /* Set up a new main loop */
    if (!(m = pa_mainloop_new())) {
        fprintf(stderr, "pa_mainloop_new() failed.\n");
        goto quit;
    }

    mainloop_api = pa_mainloop_get_api(m);

    r = pa_signal_init(mainloop_api);
    assert(r == 0);
    pa_signal_new(SIGINT, exit_signal_callback, NULL);
#ifdef SIGPIPE
    signal(SIGPIPE, SIG_IGN);
#endif

    /* Create a new connection context */
    if (!(context = pa_context_new(mainloop_api, client_name))) {
        fprintf(stderr, "pa_context_new() failed.\n");
        goto quit;
    }

    pa_context_set_state_callback(context, context_state_callback, NULL);

    /* Connect the context */
    pa_context_connect(context, server, 0, NULL);

    /* Run the main loop */
    if (pa_mainloop_run(m, &ret) < 0) {
        fprintf(stderr, "pa_mainloop_run() failed.\n");
        goto quit;
    }

quit:
    if (stream)
        pa_stream_unref(stream);

    if (context)
        pa_context_unref(context);

    if (m) {
        pa_signal_done();
        pa_mainloop_free(m);
    }

    pa_xfree(server);
    pa_xfree(device);
    pa_xfree(client_name);
    pa_xfree(stream_name);

    if (sndfile)
        sf_close(sndfile);

    return ret;
}
